### DISCUSSION TITLE
PMD Module for Salesforce Metadata

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [pmd/pmd#4877] [General]
Hello all.

I've opened this discussion for feedback, ideas, and general updates. I'm working on contributing a language module for Salesforce metadata. I will primarily be working on this off-hours (weekends, etc.) but I have the time and the motivation.

The goal here is not to cover every possible type at the outset; rather, I want to start with the obvious: flows, and reports. Both have some easy hits (elements that require descriptions longer than n characters, etc.) As other types are brought into focus, they can be added later. 

The structure I am proposing is as follows (this is rough as to namespaces, so bear with me): 
* net.sourceforge.pmd.lang.xml.sfmetadata.flow (etc.)

I know there has been some chatter about a unified Salesforce namespace, but I wanted to get the ball rolling and get something out there for comment. I think we can unify everything when it seems feasible, but that unification comes at a cost: it will break existing code and there is wide adoption in that community.

Please share your thoughts, ideas, snide remarks, etc.! I'm excited to put something together. I've already hacked together some XML-based rules for a current project, so I'm comfortable working on something more permanent. 

Incidentally this would put PMD on par (as it's own, standalone solution) with Autorabit's Codescan (at least as far as I can tell.)

@jsotuyod Please advise if this description needs some work.

A happy PMD user,
Justin Stroud

**Comment by justinstroud-campmeeker** [comment]
First thing's first...I'm on the fence as to the namespace. It's going to be either:
- net.sourceforge.pmd.lang.xml.sfmetadata.<metadata types here, flows,reports, etc.>.*
- net.sourceforge.pmd.lang.salesforce.metadata.<flows,reports, etc.>

The second option ties in with the unified salesforce namespace (see #3846.) The first, well, I'm not crazy about using a platform acronym in a namespace. It goes against my nature to be even a little opaque in that regard.

I would love input on this sooner than later, as it's going to inform my first steps in creating this module. Tagging @codefriar as I think he's interested in following this.

The old quote applies here:

"The two hardest things in programming are cache invalidation and naming things." - ?

**Comment by jsotuyod** [comment]
> The structure I am proposing is as follows (this is rough as to namespaces, so bear with me):
> 
> net.sourceforge.pmd.lang.xml.sfmetadata.flow (etc.)

I'm assuming this will be under the `pmd-xml` module, right? It's the only module currently hosting more than one language, with support for raw XML, POMs, etc.

> The second option ties in with the unified salesforce namespace (see https://github.com/pmd/pmd/issues/3846.) The first, well, I'm not crazy about using a platform acronym in a namespace. It goes against my nature to be even a little opaque in that regard.

I'd not do this at the time being. Moving Apex / Visualforce to a different artifact to get them all together would be a breaking change (big no-no for non-major releases). Moreover, if we wanted to do something like this, we could use a meta-module (ie: a pom shipping `pmd-apex`, `pmd-visualforce` and `pmd-xml` bundled together, and include something similar to Java's `quickstart` ruleset for these languages combined. It's definitely more of a distribution issue than an issue of where the code lays.

**Reply by justinstroud-campmeeker** [reply]
> I'm assuming this will be under the pmd-xml module, right? It's the only module currently hosting more than one language, with support for raw XML, POMs, etc.

Makes sense. So, the namespace options would be (including my 2 cents):
1. ```net.sourceforge.pmd.lang.xml.salesforce.metadata.```(flows, reports, etc.)
1. ```net.sourceforge.pmd.lang.xml.```(flows, reports, etc.)
1. ```net.sourceforge.pmd.lang.xml.sfmetadata.```(flows, reports, etc.)

I really dislike the idea of an acronym, and I'm also unsure about placing ```flow```, ```report```, etc. directly under ```xml``` because "flow" and "report" are pretty generic terms. Dealer's choice, I'll go whatever direction you prefer.

Yeah, I think I mentioned the issue with doing the unified thing at this point for the very reasons you mentioned. Just kind of a bad idea to throw in a breaking change (especially right after a major release, right?)

I already forked the repo and created a branch off of that master...it seemed the most logical thing to do.

Thanks!

**Reply by adangel** [reply]
I'm not sure, how far you are with your experiments. But one different way would be, to develop a separate module completely _outside_ of the PMD source tree. That is, you add pmd-core and pmd-xml as dependencies. Once this is working, we can think about, how and where we want to place it inside PMD. Just a suggestion from my side.

Regarding the namespace question: If it goes into the module "pmd-xml", then it must be prefixed with `net.sourceforge.pmd.lang.xml`. Everything after that doesn't matter that much, as long as it "makes sense" (whatever that means). I would prefer something that as "sf" or "salesforce" in it. Just using "flow" or "reports" sounds a bit too generic. The flow also as a XML namespace, doesn't it? Maybe we can take/reuse some parts of that name?

> I think we can unify everything when it seems feasible, but that unification comes at a cost: it will break existing code and there is wide adoption in that community.

Yes, we should do it step by step and not unify something that doesn't exist yet. Let's first have the pieces that we want to unify. Only then, we can see, how and if it makes sense to unify this at all. I agree with what @jsotuyod  said: "It's more a distribution issue".

**Reply by justinstroud-campmeeker** [reply]
I think pulling core and xml into a separate module and working from there is sound advice, thanks.

**Reply by adangel** [reply]
This will of course create a bit more effort, when we want to integrate it into the main PMD repository. But it's easier, to try and develop rules that way.

I came to my mind, that you essentially are creating custom rules. We have this e.g. https://github.com/pmd/pmd-examples - which gives a starting point for writing custom rules for java. But it should be easily adaptable for custom rules for XML. You additionally get the ability, to create a custom distribution which contains PMD + your custom rule - this might make it easier for any potential users to use it right away.

**Reply by justinstroud-campmeeker** [reply]
I was thinking about that as the examples repo is pretty bare bones right now, and I'd love to flesh that out further for the project.

> I came to my mind, that you essentially are creating custom rules.  

I see what you are saying about creating custom rules. In a sense, that is true. Howewer...

The one thing that is sticking in my craw though is that the way Salesforce metadata XML is constructed is very domain specific. Some things define schemas (custom fields, custom objects), some things define access control (which is essentially a schema with a different suit on), and some things define behavior (like flows...it's not uncommon to have the formula meta-language in there).

What would truly add value to the community would be a language module that treats all of those things in a systematic, formal manner. Meaning I almost think that flows (for example, as a first pass) could benefit with the structure of an AST-based visitor traversal. With all that comes certain benefits:

- Well-typed AST nodes for things like header information, variable and constant definitions
- Clear delineations of node types (DML operations versus logical operations)
- Creating distinctive rules because of the above parameters becomes easier

I also see the hard part:

- New language module becomes more complicated
- Because of the above point, it requires further work and review

Flows would be a good place to start, because it's pretty easy to hack together XML rules for some of the other objects that are less "computational" tasks and more schema definitions. 
I'm pretty confident that Salesforce isn't going to up and change their flow schema tomorrow (or in the next year) because that would break a ton of existing implementations. This schema is well documented and stable. It's more likely to be added on to than changed outright.

> This will of course create a bit more effort, when we want to integrate it into the main PMD repository. But it's easier, to try and develop rules that way.

And if I'm not mistaken, my proposal to create an true AST-vistor-based module for (at the beginning) flows makes it even more complicated :)

Does that make sense? Sorry for the wall of text! Trying to make sure I'm being as clear as possible.

Justin

**Reply by justinstroud-campmeeker** [reply]
Replying to myself here, but I was thinking...

>  Some things define schemas (custom fields, custom objects), some things define access control (which is essentially a schema with a different suit on), and some things define behavior (like flows...it's not uncommon to have the formula meta-language in there).

In the spirit of keeping PMD (at its core) as objective as possible (assuming that is one of the goals of the overall project), it occurs to me that the domain-specific nature of Salesforce Metadata almost _cries out_ for an independent module.

**Comment by adangel** [comment]
Collecting already existing issues, that are related here:

* #3846
* #3849
* #3952
* #3413
* https://github.com/pmd/pmd-github-action/issues/265

**Comment by justinstroud-campmeeker** [comment]
Hey @adangel and @jsotuyod . I'm still looking at this as a contribution. I've been busy with work, but much of that will inform me moving forward. Most importantly, I've had the opportunity to work with PMD 7 in the wild and leverage node streams in an HTML context. 

I'm starting to think that the way the HTML (or even VisualForce) language module is built might better reflect the nuances of Salesforce metadata. Although processing XML-as-XML is great, I find the context switching between the AST structure and the DOM method to be cumbersome compared to a unified approach.

I've always leaned towards consistency, so that's where I get that from.
